lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+                    params, log, cb);
+            } else if (params && params.versionId && !params.repairMaster) {
+                return this.putObjectVerCase3(c, bucketName, objName, objVal,
+                    params, log, cb);
+            } else if (params && params.versionId && params.repairMaster) {
+                return this.putObjectVerCase4(c, bucketName, objName, objVal,
+                    params, log, cb);
+            }
+            return this.putObjectNoVer(c, bucketName, objName, objVal,
                 params, log, cb);
-        }
-        return this.putObjectNoVer(c, bucketName, objName, objVal,
-            params, log, cb);
+        });
     }
 
+    /**
+     * gets versionned and non versionned object metadata
+     * @param {String} bucketName bucket name
+     * @param {String} objName object name
+     * @param {object} params params
+     * @param {String} params.versionId version of object (optional)
+     * @param {Object} log logger
+     * @param {Function} cb callback
+     * @return {undefined}
+     */
     getObject(bucketName, objName, params, log, cb) {
         const c = this.getCollection(bucketName);
-        if (params && params.versionId) {
-            // eslint-disable-next-line
-            objName = formatVersionKey(objName, params.versionId);
-        }
-        c.findOne({
-            _id: objName,
-        }, {}, (err, doc) => {
-            if (err) {
-                log.error('findOne: error getting object',
-                    { error: err.message });
-                return cb(errors.InternalError);
-            }
-            if (!doc) {
-                return cb(errors.NoSuchKey);
-            }
-            if (doc.value.isPHD) {
-                this.getLatestVersion(c, objName, log, (err, value) => {
+        let key;
+        async.waterfall([
+            next => this.getBucketVFormat(bucketName, log, next),
+            (vFormat, next) => {
+                if (params && params.versionId) {
+                    // eslint-disable-next-line
+                    key = formatVersionKey(objName, params.versionId, vFormat);
+                } else {
+                    // eslint-disable-next-line
+                    key = formatMasterKey(objName, vFormat);
+                }
+                c.findOne({
+                    _id: key,
+                }, {}, (err, doc) => {
                     if (err) {
-                        log.error('getLatestVersion: getting latest version',
+                        log.error('findOne: error getting object',
                             { error: err.message });
-                        return cb(err);
+                        return next(errors.InternalError);
                     }
-                    return cb(null, value);
+                    if (!doc) {
+                        return next(errors.NoSuchKey);
+                    }
+                    if (doc.value.isPHD) {
```

Comment:
I think we could do `return next(null, doc)` and create a new step in the waterfall for the following logic